### PR TITLE
[GOVCMSD8-556]: Add read replica access.

### PR DESCRIPTION
**Pull request by steveworley**
- Updates `govcms-deploy` to use the read replica for
- Adds database config for the read replica
- Introduces `MARIADB_HOST_READREPLICA` environment variable to control the mysql host.

https://github.com/amazeeio/dbaas-mariadb-apb/pull/5#issuecomment-551346321

### CODE REVIEW

File: .docker/images/govcms8/settings/settings.php [hunk 1]
```diff
@@ -118,6 +118,20 @@ public function __toString() {
     'charset' => 'utf8mb4',
     'collation' => 'utf8mb4_general_ci',
   ];
+
+  if (getenv('MARIADB_HOST_READREPLICA')) {
```
Comment (srowlands) [inline .docker/images/govcms8/settings/settings.php]: To confirm: naming in Lagoon (`MARIADB_READREPLICA_HOST`)

File: .docker/images/govcms8/settings/settings.php [hunk 2]
```diff
@@ -118,6 +118,20 @@ public function __toString() {
...
+    $databases['read']['default'] = [
+      'driver' => 'mysql',
+      'database' => getenv('MARIADB_DATABASE') ?: 'drupal',
+      'username' => getenv('MARIADB_USERNAME') ?: 'drupal',
+      'password' => getenv('MARIADB_PASSWORD') ?: 'drupal',
+      'host' => getenv('MARIADB_HOST_READREPLICA') ?: 'mariadb',
+      'host' => getenv('MARIADB_HOST') ?: 'mariadb',
```
Comment (srowlands) [inline .docker/images/govcms8/settings/settings.php]: Overriding rr

**Comment by steveworley** [comment]
Updated the PR to include the replica configuration for the default database service, this allows us to use the `database.replica` service in custom code to send requests directly to the reader where necessary.

Drupal will round robin all configured replicas when it is making this connection.

File: .docker/images/govcms8/scripts/govcms-deploy [hunk 3]
```diff
@@ -41,19 +50,21 @@ common_deploy () {
...
 if [[ "$LAGOON_ENVIRONMENT_TYPE" != "production" ]]; then
+
+    read_replica_enabled $sync_opts
+
     if ! drush status --fields=bootstrap | grep -q "Successful"; then
         # Import prod db in Lagoon development environments.
         if [[ ! -z "$LAGOON_ENVIRONMENT_TYPE" && "$LAGOON_ENVIRONMENT_TYPE" != "local" ]]; then
-          drush sql-sync @govcms.prod @self -y
-          common_deploy
+          drush sql-sync @govcms-prod @self $sync_opts -y
```
Comment (srowlands) [inline .docker/images/govcms8/scripts/govcms-deploy]:
```suggestion
          drush sql-sync @govcms-prod @self $sync_opts -y
          common_deploy
```

Comment (srowlands) [reply .docker/images/govcms8/scripts/govcms-deploy]: Think you accidentally removed common_deploy -- still want updb+cim to run after syncing from r/o db

File: .docker/images/govcms8/scripts/govcms-deploy [hunk 4]
```diff
@@ -20,6 +20,15 @@ mkdir -p /app/web/sites/default/files/private/tmp/
 config_count=`ls -1 /app/config/default/*.yml 2>/dev/null | wc -l`
 dev_config_count=`ls -1 /app/config/dev/*.yml 2>/dev/null | wc -l`
 
+# Database options to configure the remote to use the read replica when
+# performing read operations.
+sync_opts=""
+read_replica_enabled () {
+  if tables=$(drush @govcms-prod sqlq 'show tables; --database=read') && [ -n "$tables" ]; then
```
Comment (srowlands) [inline .docker/images/govcms8/scripts/govcms-deploy]:
```suggestion
  if tables=$(drush @govcms-prod sqlq 'show tables;' --database=read) && [ -n "$tables" ]; then
```

File: .docker/images/govcms8/settings/settings.php [hunk 5]
```diff
@@ -118,6 +118,37 @@ public function __toString() {
...
   ];
+
+  if (getenv('MARIADB_READREPLICA_HOST')) {
+    
+    // Add replica support to the default database connection. This allows services
+    // to use the database.replica service for particular operations.
+    // @TODO: Lagoon should expose MARAIDB replica hosts as an array so we can
+    // scale the replicas horizontally.
+    $database['default']['replica'][] = [
+      'drive' => 'myslq',
```
Comment (srowlands) [inline .docker/images/govcms8/settings/settings.php]:
```suggestion
      'driver' => 'mysql', 
```

File: .docker/images/govcms8/scripts/govcms-deploy (hunk 4 above)
Comment (srowlands) [reply .docker/images/govcms8/scripts/govcms-deploy]: Bail early if env var is not present; prevent a failed bootstrap.

Redirect stderr away, we do not need to see failures here (but should prob echo out so we know which db is being used for sync)

File: .docker/images/govcms8/scripts/govcms-deploy [hunk 6]
```diff
@@ -41,10 +50,13 @@ common_deploy () {
...
 # Non production environments.
 if [[ "$LAGOON_ENVIRONMENT_TYPE" != "production" ]]; then
+
+    read_replica_enabled $sync_opts
+
     if ! drush status --fields=bootstrap | grep -q "Successful"; then
         # Import prod db in Lagoon development environments.
         if [[ ! -z "$LAGOON_ENVIRONMENT_TYPE" && "$LAGOON_ENVIRONMENT_TYPE" != "local" ]]; then
-          drush sql-sync @govcms.prod @self -y
+          drush sql-sync @govcms-prod @self $sync_opts -y
```
Comment (srowlands) [inline .docker/images/govcms8/scripts/govcms-deploy]:
```suggestion
          drush sql-sync @govcms.prod @self $sync_opts -y
```

Comment (srowlands) [reply .docker/images/govcms8/scripts/govcms-deploy]: This shouldn't change -- d8 standard drush alias is `govcms.prod`

File: .docker/images/govcms8/scripts/govcms-deploy [hunk 7]
```diff
@@ -53,7 +65,7 @@ if [[ "$LAGOON_ENVIRONMENT_TYPE" != "production" ]]; then
       # @see comments for GOVCMS_TEST_CANARY above.
       if [[ "$GOVCMS_TEST_CANARY" = TRUE ]]; then
         echo "GOVCMS_TEST_CANARY is set, syncing the database."
-        drush sql-sync @govcms.prod @self -y
+        drush sql-sync @govcms-prod @self $sync_opts -y
```
Comment (srowlands) [inline .docker/images/govcms8/scripts/govcms-deploy]:
```suggestion
        drush sql-sync @govcms.prod @self $sync_opts -y
```

**Comment by AlexSkrypnyk** [comment]
@srowlands could you please re-review this one. thanks

**Comment by steveworley** [comment]
@AlexSkrypnyk as far as I know the Lagoon broker will be exposing it as `MARIADB_READREPLICA_HOSTS` https://github.com/amazeeio/dbaas-mariadb-apb/pull/5/files#diff-e349823b467059e175ddb237223092ecR50

**Comment by AlexSkrypnyk** [comment]
I have cross-checked requested changes by @srowlands with the latest changes in code - all points were addressed. This is ready to be merged.

**Comment by AlexSkrypnyk** [comment]
Similar PR is coming for `scaffold-tooling`.

**Comment by AlexSkrypnyk** [comment]
@steveworley this is ready to be merged. please merge at your earliest convenience

**Pull request merged**